### PR TITLE
[PVR] Guide window: Fix selection after channelgroup switching

### PR DESCRIPTION
**Pull request by ksooo**
This PR fixes wrong epg event selection after channel group switching in the PVR Guide window. Expected behavior is that the same event is still selected after a channel group switch, if the previously selected channel is also part of the new channel group. Otherwise, the first channel of the new group is selected.

The second commit introduces asynchrounous channel group switching. The gui render thread is no longer blocked while the epg grid gets refilled. A wait dialog is shown if the refill lasts longer than 100 msecs.

This was runtime tested on macOS and Linux on latest Kodi master.

@Jalle19 for code review?
@xhaggi for runtime testing (you reported the problems ;-)

### CODE REVIEW

**Comment by ksooo** [comment]
Just discovered that this us not yet ready to merge as I'm now doing a little too aggressive caching. Needs some more love... Adding WIP batch for now.

**Comment by Jalle19** [comment]
Ping me again when you want a new review.

**Comment by xhaggi** [comment]
same here, let me know if you're happy with it.

**Comment by ksooo** [comment]
@Jalle19 @xhaggi Ping. :-) The PR imo is ready for review and runtime test.

**Comment by ksooo** [comment]
@MilhouseVH could you please include this PR into your next build, to get early feedback in case I broke stuff.

**Comment by MilhouseVH** [comment]
Will do.

File: xbmc/pvr/windows/GUIWindowPVRGuide.h [hunk 1]
```diff
@@ -79,7 +80,7 @@ namespace PVR
     void StartRefreshTimelineItemsThread();
     void StopRefreshTimelineItemsThread();
 
-    std::unique_ptr<CPVRRefreshTimelineItemsThread> m_refreshTimelineItemsThread;
+    std::shared_ptr<CPVRRefreshTimelineItemsThread> m_refreshTimelineItemsThread;
```
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRGuide.h]: why did you change this?

File: xbmc/pvr/windows/GUIWindowPVRGuide.cpp [hunk 2]
```diff
@@ -193,27 +205,21 @@ bool CGUIWindowPVRGuide::Update(const std::string &strDirectory, bool updateFilt
...
+      refreshTimelineItemsThread = m_refreshTimelineItemsThread;
     }
   }
 
-  // never call RefreshTimelineItems with locked mutex!
-  if (bRefresh)
-    RefreshTimelineItems();
+  // never call DoRefresh with locked mutex!
+  if (refreshTimelineItemsThread)
+    refreshTimelineItemsThread->DoRefresh();
```
Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: if you don't use a copy of the ptr and leave the boolean to indicate the refresh you don't need to make it a shared_pointer.

Comment (ksooo) [inline xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: I have to ensure that the thread object gets not destroyed while I'm not holding the lock! The point was not to get rid of the bool. This was just a side effect.

Comment (xhaggi) [inline xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: how could this be done, if CGUIWindowPVRGuide own the thread?

Comment (ksooo) [inline xbmc/pvr/windows/GUIWindowPVRGuide.cpp]: Very good point. Currently the methods using the thread object and the method that creates/destroys the thread object are exclusively called by the gui thread. I will change the code. Thx.

**Comment by ksooo** [comment]
@xhaggi the changes you requested are now in.

**Comment by xhaggi** [comment]
Looks good thanks

**Review by Jalle19 — APPROVED** [review]

**Comment by ksooo** [comment]
Thanks guys. I guess I will wait with the merge for some days to give milhouse build users the chance for some testing.

**Comment by ksooo** [comment]
In case we will do a 17.2 release, should we backport this?

**Comment by xhaggi** [comment]
Sure, because it's a really annoying bug.

**Comment by ksooo** [comment]
rebased

jenkins build this please

**Comment by Jalle19** [comment]
+1 for backport if you're up for it

**Comment by ksooo** [comment]
jenkins  build this please

**Comment by ksooo** [comment]
jenkins build this please
